File: app/src/main/java/com/mobile/hinde/spacetime/Act_Image.java
package com.mobile.hinde.spacetime;

import android.content.Intent;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.util.DisplayMetrics;
import android.util.Log;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.SetOptions;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.mobile.hinde.connection.AsyncResponse;
import com.mobile.hinde.connection.Image_List;
import com.mobile.hinde.connection.Image_URL;
import com.mobile.hinde.utils.Tool;
import com.mobile.hinde.utils.UserSettings;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class Act_Image extends AppCompatActivity {

    private ImageView mImage;
    private TextView mLegend;
    private Button mSave;

    private String mImageName;
    private String mImageTitle;
    private String mImageLegend;
    private String mURL;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_image);

        DisplayMetrics dm = new DisplayMetrics();
        getWindowManager().getDefaultDisplay().getMetrics(dm);

        final int mMaxWidth = (int)(dm.widthPixels * 0.8);
        final int mMaxHeight = (int)(dm.heightPixels * 0.8);

        final TypedArray styledAttributes = getTheme().obtainStyledAttributes(
                new int[] { android.R.attr.actionBarSize }
        );
        final int actionBarHeight = (int) styledAttributes.getDimension(0, 0);
        styledAttributes.recycle();

        mLegend = findViewById(R.id.text_legend);
        mSave = findViewById(R.id.save);
        mImage = findViewById(R.id.image);


        Intent intent = getIntent();
        mImageName = intent.getExtras().getString("name");
        mImageTitle = intent.getExtras().getString("title");
        mImageLegend = intent.getExtras().getString("legend");
        mURL = intent.getExtras().getString("URL");

        new Image_URL(new AsyncResponse(){
            @Override
            public void processFinish(Object output) {        // Create a storage reference from our app
                mImage.setImageDrawable((Drawable) output);
                int width = ((Drawable)output).getIntrinsicWidth();
                int height = ((Drawable)output).getIntrinsicHeight();
                if(width > 0) {
                    float ratio = (float) mMaxWidth / (float) width;
                    int newHeight = (int) (height * ratio) + mLegend.getHeight() + mSave.getHeight() + actionBarHeight;
                    getWindow().setLayout(mMaxWidth, newHeight);
                }
            }
        }).execute(mURL);

        getSupportActionBar().setTitle(mImageTitle);
        mLegend.setText(mImageLegend);
        getWindow().setLayout(mMaxWidth, mMaxHeight);
    }

    @Override
    public void onResume() {
        super.onResume();
    }

    public void onDestroy(){
        super.onDestroy();
    }

}

File: app/src/main/java/com/mobile/hinde/spacetime/ActImage.kt
package com.mobile.hinde.spacetime

import android.graphics.BitmapFactory
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.DisplayMetrics
import android.widget.ImageView
import android.widget.TextView
import com.mobile.hinde.connection.AsyncResponse
import com.mobile.hinde.connection.ImageURL
import com.mobile.hinde.utils.Tools


class ActImage : AppCompatActivity(){

    private var mImage: ImageView? = null
    private var mLegend: TextView? = null

    private var mImageName: String? = null
    private var mImageTitle: String? = null
    private var mImageLegend: String? = null
    private var mURL: String? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_image)

        val dm = DisplayMetrics()
        windowManager.defaultDisplay.getMetrics(dm)

        val mMaxWidth = (dm.widthPixels * 0.8).toInt()
        val mMaxHeight = (dm.heightPixels * 0.8).toInt()

        val styledAttributes = theme.obtainStyledAttributes(
                intArrayOf(android.R.attr.actionBarSize)
        )
        val actionBarHeight = styledAttributes.getDimension(0, 0f).toInt()
        styledAttributes.recycle()

        window.setLayout(mMaxWidth, mMaxHeight)

        mLegend = findViewById(R.id.text_legend)
        mImage = findViewById(R.id.image)


        val intent = intent
        mImageName = intent.extras!!.getString("name")
        mImageTitle = intent.extras!!.getString("title")
        mImageLegend = intent.extras!!.getString("legend")
        mURL = intent.extras!!.getString("URL")


        supportActionBar!!.title = mImageTitle
        mLegend!!.text = mImageLegend
        mLegend!!.measure(0,0)

        if(fileList().contains(mImageName)){
            val bmpImage = BitmapFactory.decodeStream(openFileInput(mImageName))
            mImage!!.setImageBitmap(bmpImage)
            val width = bmpImage.width
            val height = bmpImage.height
            if (width > 0) {
                val ratio = mMaxWidth.toFloat() / width.toFloat()
                val newHeight = (height * ratio).toInt() + mLegend!!.measuredHeight + actionBarHeight
                window.setLayout(mMaxWidth, newHeight)
            }
        }else{
            ImageURL(object : AsyncResponse {
                override fun processFinish(output: Any) {
                    // Create a storage reference from our app
                    if(output == null){
                        Tools.displayError(this@ActImage , resources.getString(R.string.error_title), resources.getString(R.string.error_image_not_found))
                    }else {
                        mImage!!.setImageDrawable(output as Drawable)
                        val width = output.intrinsicWidth
                        val height = output.intrinsicHeight
                        if (width > 0) {
                            val ratio = mMaxWidth.toFloat() / width.toFloat()
                            val newHeight = (height * ratio).toInt() + mLegend!!.height + actionBarHeight
                            window.setLayout(mMaxWidth, newHeight)
                        }
                    }
                }
            }).execute(mImageName, mURL)
        }

    }


}
